File: 2023/day05/python/solution.py
"""Advent of Code 2023 - Day 5: If You Give A Seed A Fertilizer

This module solves the almanac seed-to-location mapping puzzle. Seeds must be
mapped through a series of transformation maps (seed->soil->fertilizer->etc.)
to find their final location numbers.

Part 1: Find the minimum location for a list of individual seed numbers.
Part 2: Interpret seeds as ranges and find the minimum location efficiently
        using interval arithmetic rather than brute force iteration.
"""

from dataclasses import dataclass
from pathlib import Path


@dataclass(frozen=True, slots=True)
class MapRange:
    """A single mapping range that transforms source values to destination values.

    Attributes:
        destination_start: The starting value of the destination range.
        source_start: The starting value of the source range.
        length: The number of values in the range.
    """
    destination_start: int
    source_start: int
    length: int

    @property
    def source_end(self) -> int:
        """Return the exclusive end of the source range."""
        return self.source_start + self.length

    @property
    def offset(self) -> int:
        """Return the offset to add when mapping source to destination."""
        return self.destination_start - self.source_start

    def contains(self, value: int) -> bool:
        """Check if a value falls within the source range."""
        return self.source_start <= value < self.source_end

    def transform(self, value: int) -> int:
        """Transform a source value to its destination value.

        Assumes the value is within the source range.
        """
        return value + self.offset


@dataclass(frozen=True, slots=True)
class Interval:
    """A half-open interval [start, end) representing a range of values.

    Attributes:
        start: The inclusive start of the interval.
        end: The exclusive end of the interval.
    """
    start: int
    end: int

    def is_valid(self) -> bool:
        """Check if this interval is non-empty."""
        return self.start < self.end

    def intersect(self, other: 'Interval') -> 'Interval':
        """Return the intersection of two intervals."""
        return Interval(max(self.start, other.start), min(self.end, other.end))


def parse_input(text: str) -> tuple[list[int], list[list[MapRange]]]:
    """Parse the puzzle input into seeds and transformation maps.

    Args:
        text: The raw puzzle input text.

    Returns:
        A tuple containing:
        - A list of seed numbers
        - A list of maps, where each map is a list of MapRange objects
    """
    sections = text.strip().split('\n\n')

    # Parse seeds from the first line
    seeds = list(map(int, sections[0].split(': ')[1].split()))

    # Parse each transformation map
    maps: list[list[MapRange]] = []
    for section in sections[1:]:
        lines = section.strip().split('\n')
        ranges: list[MapRange] = []
        for line in lines[1:]:  # Skip the header line
            destination_start, source_start, length = map(int, line.split())
            ranges.append(MapRange(destination_start, source_start, length))
        maps.append(ranges)

    return seeds, maps


def apply_map(value: int, map_ranges: list[MapRange]) -> int:
    """Apply a transformation map to a single value.

    Searches through the map ranges to find one that contains the value.
    If found, transforms the value; otherwise returns it unchanged.

    Args:
        value: The input value to transform.
        map_ranges: The list of MapRange objects defining the transformation.

    Returns:
        The transformed value, or the original if no range matches.
    """
    for map_range in map_ranges:
        if map_range.contains(value):
            return map_range.transform(value)
    return value


def seed_to_location(seed: int, maps: list[list[MapRange]]) -> int:
    """Convert a seed number through all maps to get its location.

    Applies each transformation map in sequence, passing the result
    of each transformation to the next map.

    Args:
        seed: The initial seed number.
        maps: The list of transformation maps to apply in order.

    Returns:
        The final location number for this seed.
    """
    value = seed
    for map_ranges in maps:
        value = apply_map(value, map_ranges)
    return value


def part1(seeds: list[int], maps: list[list[MapRange]]) -> int:
    """Solve Part 1: Find the minimum location for individual seeds.

    Args:
        seeds: List of individual seed numbers.
        maps: List of transformation maps.

    Returns:
        The lowest location number among all seeds.
    """
    return min(seed_to_location(seed, maps) for seed in seeds)


def apply_map_to_intervals(
    input_intervals: list[Interval],
    map_ranges: list[MapRange]
) -> list[Interval]:
    """Apply a transformation map to a list of intervals.

    This is the key optimization for Part 2. Instead of checking each
    individual seed value, we transform entire intervals at once.

    Each input interval may be split into multiple output intervals:
    - Parts that overlap with a MapRange are transformed
    - Parts that don't overlap with any MapRange pass through unchanged

    Args:
        input_intervals: List of input intervals to transform.
        map_ranges: The transformation map to apply.

    Returns:
        List of transformed output intervals.
    """
    result: list[Interval] = []

    for interval in input_intervals:
        # Track parts of the interval not yet mapped
        unmapped: list[Interval] = [interval]

        for map_range in map_ranges:
            source_interval = Interval(map_range.source_start, map_range.source_end)
            next_unmapped: list[Interval] = []

            for remaining in unmapped:
                # Check for overlap with this map range
                overlap = remaining.intersect(source_interval)

                if overlap.is_valid():
                    # The overlapping portion gets transformed
                    transformed = Interval(
                        overlap.start + map_range.offset,
                        overlap.end + map_range.offset
                    )
                    result.append(transformed)

                    # The non-overlapping portions remain unmapped
                    before = Interval(remaining.start, overlap.start)
                    if before.is_valid():
                        next_unmapped.append(before)

                    after = Interval(overlap.end, remaining.end)
                    if after.is_valid():
                        next_unmapped.append(after)
                else:
                    # No overlap, this portion remains unmapped
                    next_unmapped.append(remaining)

            unmapped = next_unmapped

        # Any remaining unmapped portions pass through unchanged
        result.extend(unmapped)

    return result


def part2(seeds: list[int], maps: list[list[MapRange]]) -> int:
    """Solve Part 2: Find the minimum location for seed ranges.

    In Part 2, the seeds list is interpreted as pairs of (start, length),
    defining ranges of seed values. We use interval arithmetic to
    efficiently transform these ranges without iterating over individual seeds.

    Args:
        seeds: List of seed range definitions (alternating start and length).
        maps: List of transformation maps.

    Returns:
        The lowest location number among all seeds in all ranges.
    """
    # Convert seed pairs into intervals
    intervals: list[Interval] = []
    for i in range(0, len(seeds), 2):
        seed_start, seed_length = seeds[i], seeds[i + 1]
        intervals.append(Interval(seed_start, seed_start + seed_length))

    # Apply each transformation map to the intervals
    for map_ranges in maps:
        intervals = apply_map_to_intervals(intervals, map_ranges)

    # Find the minimum start value among all resulting intervals
    return min(interval.start for interval in intervals)


def main() -> None:
    """Main entry point: parse input and solve both parts."""
    input_path = Path(__file__).parent.parent / 'input.txt'
    text = input_path.read_text()

    seeds, maps = parse_input(text)

    print('Part 1:', part1(seeds, maps))
    print('Part 2:', part2(seeds, maps))


if __name__ == '__main__':
    main()

File: 2023/day05/clojure/solution.clj
(ns solution
  "Advent of Code 2023 Day 5: If You Give A Seed A Fertilizer

   Transforms seed numbers through a series of category mappings
   to find the lowest location number."
  (:require [clojure.string :as str]))

(def input (slurp "../input.txt"))

(defn parse-seeds
  "Parse the seeds line into a sequence of seed numbers."
  [line]
  (->> line
       (re-seq #"\d+")
       (map parse-long)))

(defn parse-map-line
  "Parse a single map line into [dest-start src-start length] vector."
  [line]
  (->> line
       str/trim
       (re-seq #"\d+")
       (mapv parse-long)))

(defn parse-map-section
  "Parse a map section into a vector of range mappings.
   Each mapping is [dest-start src-start length]."
  [section]
  (->> section
       str/split-lines
       rest
       (mapv parse-map-line)))

(defn parse-input
  "Parse the full input text into [seeds maps].

   Returns a vector where:
   - seeds is a sequence of seed numbers
   - maps is a vector of mapping tables, each containing range mappings"
  [text]
  (let [[seeds-section & map-sections] (-> text str/trim (str/split #"\n\n"))]
    [(parse-seeds seeds-section)
     (mapv parse-map-section map-sections)]))

(defn find-mapping
  "Find the applicable mapping for a value, if any.
   Returns the matching [dest-start src-start length] or nil."
  [value ranges]
  (some (fn [[_ src-start length :as mapping]]
          (when (and (>= value src-start)
                     (< value (+ src-start length)))
            mapping))
        ranges))

(defn apply-map-value
  "Apply a single map to transform a value.
   If no mapping matches, returns the value unchanged."
  [value ranges]
  (if-let [[dest-start src-start _] (find-mapping value ranges)]
    (+ dest-start (- value src-start))
    value))

(defn seed-to-location
  "Convert a seed number to its final location through all maps."
  [seed maps]
  (reduce apply-map-value seed maps))

(defn part1
  "Find the lowest location number for any initial seed."
  [seeds maps]
  (transduce (map #(seed-to-location % maps))
             min
             Long/MAX_VALUE
             seeds))

(defn split-range-by-mapping
  "Split an input range by a single mapping range.

   Returns a map with:
   - :mapped - the portion that was mapped (or nil)
   - :unmapped - portions that remain unmapped (may be empty)"
  [[range-start range-end] [dest-start src-start length]]
  (let [src-end (+ src-start length)
        offset  (- dest-start src-start)]
    {:mapped   (let [overlap-start (max range-start src-start)
                     overlap-end   (min range-end src-end)]
                 (when (< overlap-start overlap-end)
                   [(+ overlap-start offset) (+ overlap-end offset)]))
     :unmapped (cond-> []
                 ;; Portion before the mapping
                 (< range-start src-start)
                 (conj [range-start (min range-end src-start)])
                 ;; Portion after the mapping
                 (> range-end src-end)
                 (conj [(max range-start src-end) range-end]))}))

(defn process-range-through-mappings
  "Process a single input range through all mappings in a map section.
   Returns [mapped-ranges unmapped-ranges]."
  [input-range map-ranges]
  (loop [[current-mapping & remaining-mappings] map-ranges
         pending-ranges [input-range]
         mapped-results []]
    (if (or (nil? current-mapping) (empty? pending-ranges))
      [mapped-results pending-ranges]
      (let [split-results (map #(split-range-by-mapping % current-mapping) pending-ranges)
            newly-mapped  (into mapped-results (keep :mapped split-results))
            still-pending (into [] (mapcat :unmapped) split-results)]
        (recur remaining-mappings still-pending newly-mapped)))))

(defn apply-map-to-ranges
  "Apply a mapping table to a collection of ranges.
   Unmapped portions pass through unchanged."
  [input-ranges map-ranges]
  (reduce (fn [acc input-range]
            (let [[mapped unmapped] (process-range-through-mappings input-range map-ranges)]
              (into acc (concat mapped unmapped))))
          []
          input-ranges))

(defn seeds->ranges
  "Convert seed pairs [start length ...] into ranges [[start end] ...]."
  [seeds]
  (->> seeds
       (partition 2)
       (mapv (fn [[start length]] [start (+ start length)]))))

(defn part2
  "Find the lowest location number considering seeds as ranges.

   Seeds are interpreted as pairs [start length], defining ranges
   of consecutive seed numbers."
  [seeds maps]
  (let [seed-ranges (seeds->ranges seeds)
        final-ranges (reduce apply-map-to-ranges seed-ranges maps)]
    (transduce (map first) min Long/MAX_VALUE final-ranges)))

;; Main execution
(let [[seeds maps] (parse-input input)]
  (println "Part 1:" (part1 seeds maps))
  (println "Part 2:" (part2 seeds maps)))
